src/native/mqtt_connection.c

```diff
@@ -82,25 +83,20 @@ static void s_mqtt_jni_connection_release(struct mqtt_jni_connection *connection
     size_t old_value = aws_atomic_fetch_sub(&connection->ref_count, 1);
 
     AWS_LOGF_DEBUG(AWS_LS_MQTT_CLIENT, "mqtt_jni_connection release, ref count now = %d", (int)old_value - 1);
+}
 
-    if (old_value == 1) {
+/* The destroy function is called on Java MqttClientConnection resource release. */
+static void s_mqtt_jni_connection_destory(struct mqtt_jni_connection *connection) {
+    /* For mqtt311 client, we have to call aws_mqtt_client_connection_disconnect before releasing the underlying c
+     * connection.*/
+    if (aws_mqtt_client_connection_disconnect(
+            connection->client_connection, s_on_shutdown_disconnect_complete, connection) != AWS_OP_SUCCESS) {
 
-        /**
-         * Disable the onClosed callback, so it is not invoked on the last disconnect for clean-up.
-         * (We do not invoke any callbacks on the last disconnect during clean-up/shutdown.
-         * Calling the callback on the final disconnect will cause a segfault!)
+        /*
+         * This can happen under normal code paths if the client happens to be disconnected at cleanup/shutdown
+         * time. Log it (in case it was unexpected) and then shutdown the underlying connection manually.
          */
-        aws_mqtt_client_connection_set_connection_closed_handler(connection->client_connection, NULL, NULL);
-
-        if (aws_mqtt_client_connection_disconnect(
-                connection->client_connection, s_on_shutdown_disconnect_complete, connection) != AWS_OP_SUCCESS) {
-
-            /*
-             * This can happen under normal code paths if the client happens to be disconnected at cleanup/shutdown
-             * time. Log it (in case it was unexpected) and then invoke the shutdown callback manually.
-             */
-            s_on_shutdown_disconnect_complete(connection->client_connection, connection);
-        }
+        aws_mqtt_client_connection_release(connection->client_connection);
```

Comment:
This should be a call to s_on_shutdown_disconnect_complete.  While that function should only be doing this release, if you call an async api with a callback and it gives an error (thus not invoking the callback), the pattern should be to call that callback yourself rather than repeating the logic inside of it.